### PR TITLE
feat: add new `@noaignite/types` package

### PR DESCRIPTION
**Pull request by maeertin**

### CODE REVIEW

**Review by Dashice — APPROVED** [review]

File: apps/web/app/page.tsx [hunk 1]
```diff
@@ -75,7 +76,12 @@ export default function Page(): JSX.Element {
         </div>
       </div>
 
-      <Button appName="web" className={styles.button}>
+      <Button
+        // @ts-expect-error -- temp temp
+        as={Link as 'a'}
```
Comment (maeertin) [inline apps/web/app/page.tsx:81]: @Dashice Any idea here what could be done here? :)

File: packages/types/Polymorphism.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+>(
+  props: Overwrite<
+    ComponentPropsWithoutRef<Tag<TActualElement, TInternalElement>> & {
+      as?: Tag<TActualElement, TInternalElement>
+    },
+    TProps
+  >,
+) => PolymorphicElement<Tag<TActualElement, TInternalElement>, TProps>
+
+// ----- Forward Polymorph -----
```
Comment (adamsoderstrom) [inline packages/types/Polymorphism.ts]: **thought:** Could this be a bit misleading? 

We aren't really forwarding a polymorph. We're forwarding _a ref_ to a polymorph.

**Review by adamsoderstrom — COMMENTED** [review]
Great stuff! 
Leaving some thoughts!

File: apps/web/app/page.tsx (hunk 1 above)
Comment (maeertin) [reply apps/web/app/page.tsx]: As a first version we're publishing without the poly helper.

File: packages/types/Polymorphism.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+ * A polymorphic component's JSX-equivalent return type.
+ * Functionally similar to `ReactElement`.
+ */
+export type PolymorphicElement<TElement extends ElementType, TProps extends object> =
+  | ReactElement<TProps, TElement>
+  | ReactPortal
+  | null
+  | undefined
+
+// ----- Simple Polymorph -----
```
[the remark this answers is not in the document]
Comment (maeertin) [reply packages/types/Polymorphism.ts]: As a first version we're publishing without the poly helper.

File: packages/types/Polymorphism.ts (hunk 2 above)
Comment (maeertin) [reply packages/types/Polymorphism.ts]: As a first version we're publishing without the poly helper.

**Review by alexanderflink — APPROVED** [review]

**Pull request merged**